# 분류 평가 지표(혼동행렬) 계산해보기

어떤 질병 분류 모델을 총 180명의 데이터에 적용한 결과가 다음과 같았다.

| 실제값＼예측값   | 질병 예측 `1` | 정상 예측 `0` |   합계 |
| --------- | --------: | --------: | ---: |
| 실제 질병 `1` |       90명 |       35명 | 125명 |
| 실제 정상 `0` |       10명 |       45명 |  55명 |
| 합계        |      100명 |       80명 | 180명 |

## 목표: 평가 지표 계산

라이브러리를 사용하지 않고 다음 평가 지표를 파이썬 코드로 직접 계산해 주세요.

1. 정확도 Accuracy
2. 정밀도 Precision
3. 재현율 Recall
4. F1-score

계산 결과는 소수점 둘째 자리까지 출력해 주세요. (소수점 있을 시)

## 문제 1 : 혼동행렬 파이썬 코드 작성

다음 코드의 빈칸을 완성해 주세요.

```python
tp = 90
tn = 45
fp = 10
fn = 35

total = tp + tn + fp + fn

accuracy = ___________________________
precision = __________________________
recall = _____________________________
f1_score = ___________________________

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

print(f"Accuracy : {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall   : {recall:.2f}")
print(f"F1-score : {f1_score:.2f}")
```

## 문제 2. 결과 해석해보기

다음 질문 각각에 답하시오.

1. 도출된 정확도 수치는 무엇을 의미하는가?
2. 모델이 질병이라고 예측한 사람은 총 몇 명인가?
3. 도출된 정밀도 수치는 무엇을 의미하는가?
4. 실제 질병 환자는 총 몇 명인가?
5. 도출된 재현율 수치는 무엇을 의미하는가?
6. 이 모델이 놓친 실제 환자는 몇 명인가?
7. 질병 환자를 놓치지 않는 것이 중요하다면 어떤 지표를 우선 확인해야 하는가?


## 문제 1 : 혼동행렬 파이썬 코드 작성

In [1]:
# 다음 코드의 ????를 채워주세요.

# 지표별 예측결과
tp = 90
tn = 45
fp = 10
fn = 35

total = tp + tn + fp + fn

accuracy = (tp + tn) / total
precision = tp / (tp + fp)
recall = tp / (tp + fn)
f1_score = 2 * precision * recall / (precision + recall)

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

print(f"Accuracy : {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall   : {recall:.2f}")
print(f"F1-score : {f1_score:.2f}")

TP: 90
TN: 45
FP: 10
FN: 35
Accuracy : 0.75
Precision: 0.90
Recall   : 0.72
F1-score : 0.80


## 문제 2. 결과 해석해보기
**답안지 (각 질문별로 아래 답안지 작성 예시처럼 답을 추가하시고 제출해주시면 됩니다. 코드 셀, 마크다운 모두 상관없음)**
1. 도출된 정확도 수치는 무엇을 의미하는가?
2. 모델이 질병이라고 예측한 사람은 총 몇 명인가?
3. 도출된 정밀도 수치는 무엇을 의미하는가?
4. 실제 질병 환자는 총 몇 명인가?
5. 도출된 재현율 수치는 무엇을 의미하는가?
6. 이 모델이 놓친 실제 환자는 몇 명인가?
7. 질병 환자를 놓치지 않는 것이 중요하다면 어떤 지표를 우선 확인해야 하는가?

답안 작성
- 1. 정확도는 0.75(75%)로, 전체 180명 중 135명을 올바르게 예측함.
- 2. 해당 모델이 질병이라고 예측한 사람은 TP + FP = 90 + 10 = 100명
- 3. 질병이라고 예측한 100명 중 실제 질병 환자가 90명(90%)이라는 의미.
- 4. 실제 질병 환자는 TP + FN = 90 + 35 = 125명.
- 5. 재현율은 0.72(72%)로, 실제 질병 환자 125명 중 90명을 질병 환자로 올바르게 찾아냄을 의미.
- 6. 모델이 놓친 실제 환자는 FN에 해당하는 35명.
- 7. 질병 진단에서는 실제 환자를 정상이라고 판단하는 FN이 위험하므로, 질병 환자를 놓치지 않기 위해 재현율을 우선 확인해야 한다.